src/pefile.cpp

```diff
@@ -2146,16 +2162,6 @@ void PeFile::checkHeaderValues(unsigned subsystem, unsigned mask,
                      subsystem);
         throwCantPack(buf);
     }
-    if (mask & (
-             (1u<<IMAGE_SUBSYSTEM_EFI_APPLICATION)
-            |(1u<<IMAGE_SUBSYSTEM_EFI_BOOT_SERVICE_DRIVER)
-            |(1u<<IMAGE_SUBSYSTEM_EFI_RUNTIME_DRIVER)
-            |(1u<<IMAGE_SUBSYSTEM_EFI_ROM)
-              ))
-    {
-        infoWarning("PE: subsystem %u (EFI) lacks a run-time decompressor",
-            subsystem);
-    }
```

Comment:
Why/how this still exists in the file?

Reply:
This PR removed it, so should not be there any more. Note that it was added to the devel branch before this PR.

Reply:
No explicit run-time de-compressor was added for those cases, and nobody claimed an actual successful test run, so there is no evidence that it works.

Reply:
> No explicit run-time de-compressor was added for those cases, and nobody claimed an actual successful test run, so there is no evidence that it works.

I think there is, at least for `IMAGE_SUBSYSTEM_EFI_APPLICATION`. I've tested it with [shellx64](https://gitlab.com/tianocore_uefi_duet_builds/tianocore_uefi_duet_installer/-/tree/master/Shell) and [memtest86](https://www.memtest86.com), and it works.
Also, the code is not correct - it should be:
```cpp
    if ((1u << subsystem) & (
             (1u << IMAGE_SUBSYSTEM_EFI_APPLICATION)
            |(1u << IMAGE_SUBSYSTEM_EFI_BOOT_SERVICE_DRIVER)
            |(1u << IMAGE_SUBSYSTEM_EFI_RUNTIME_DRIVER)
            |(1u << IMAGE_SUBSYSTEM_EFI_ROM)
              ))
    {
        infoWarning("PE: subsystem %u (EFI) lacks a run-time decompressor",
            subsystem);
    }
```

Reply:
@bitraid, note that this code was removed by this PR, not added. I removed it when I implemented support for EFI. There is no need to fix it, because it does not exist in the current devel branch.

Reply:
@kornelpal , I know that the code was added and then removed when the PR was finished. The strange thing is that it **does** exist in devel branch.

Reply:
It was added by @jreiser in f75ad8b before I created this PR, and then I removed it in this PR. This PR was already merged, so it is not in the devel branch anymore either.

Reply:
Maybe I'm misunderstanding something, but the devel branch is now at 907c17bacdc8a946c278ad99f988d57ba0ebcb45, and the code is there:
https://github.com/upx/upx/blob/907c17bacdc8a946c278ad99f988d57ba0ebcb45/src/pefile.cpp#L2165-L2174

Reply:
@bitraid, I see, you are right.
@jreiser, by looking at the [history of pefile.cpp](https://github.com/upx/upx/commits/devel/src/pefile.cpp) in e867dda these removed lines are back, but in 4d7df85, the previous commit in the history of pefile.cpp, the lines are gone. Neither shows the diff of e867dda the re-addition of these lines. On the other hand the commit message for e867dda is "Merge branch 'devel' into jreiser-0522", so it was a merge in the opposite direction than is usually done, that might explain why the addition is not shown. Fixing this specific difference is is simple, but I don't know what else might have been changed back to a previous state in error.

Reply:
218072b2d8c04bfed54083a92e7902048d4d1147 on 2021-12-31 11:45:28 by Kornel Pal is the most recent commit in his "PE: ..." series.
e867dda8cd96f25a86df7936abf16361bff47d21 on 2022-03-16 11:30:39 by me is "Merge branch 'devel' into jreiser-0522".

The diff between those two, restircted to pefile.cpp:
```
$ git diff 218072b2d8c04bfed54083a92e7902048d4d1147 e867dda8cd96f25a86df7936abf16361bff47d21 -- pefile.cpp 
diff --git a/src/pefile.cpp b/src/pefile.cpp
index 94934d2b..076f9540 100644
--- a/src/pefile.cpp
+++ b/src/pefile.cpp
@@ -2162,6 +2162,16 @@ void PeFile::checkHeaderValues(unsigned subsystem, unsigned mask,
                      subsystem);
         throwCantPack(buf);
     }
+    if (mask & (
+             (1u<<IMAGE_SUBSYSTEM_EFI_APPLICATION)
+            |(1u<<IMAGE_SUBSYSTEM_EFI_BOOT_SERVICE_DRIVER)
+            |(1u<<IMAGE_SUBSYSTEM_EFI_RUNTIME_DRIVER)
+            |(1u<<IMAGE_SUBSYSTEM_EFI_ROM)
+              ))
+    {
+        infoWarning("PE: subsystem %u (EFI) lacks a run-time decompressor",
+            subsystem);
+    }
     //check CLR Runtime Header directory entry
     if (IDSIZE(PEDIR_COMRT))
         throwCantPack(".NET files are not yet supported");
@@ -2514,7 +2524,8 @@ void PeFile::pack0(OutputFile *fo, ht &ih, ht &oh,
     const unsigned c_len = ((ph.c_len + ic) & 15) == 0 ? ph.c_len : ph.c_len + 16 - ((ph.c_len + ic) & 15);
     obuf.clear(ph.c_len, c_len - ph.c_len);
 
-    const unsigned s1size = ALIGN_UP(ic + c_len + codesize, (unsigned) sizeof(LEXX)) + sotls + soloadconf;
+    const unsigned aligned_sotls = ALIGN_UP(sotls, (unsigned)sizeof(LEXX));
+    const unsigned s1size = ALIGN_UP(ic + c_len + codesize, (unsigned) sizeof(LEXX)) + aligned_sotls + soloadconf;
     const unsigned s1addr = (newvsize - (ic + c_len) + oam1) &~ oam1;
 
     const unsigned ncsection = (s1addr + s1size + oam1) &~ oam1;
@@ -2541,15 +2552,15 @@ void PeFile::pack0(OutputFile *fo, ht &ih, ht &oh,
     ODSIZE(PEDIR_BOUNDIM) = 0;
 
     // tls & loadconf are put into section 1
-    ic = s1addr + s1size - sotls - soloadconf;
+    ic = s1addr + s1size - aligned_sotls - soloadconf;
 
     if (use_tls_callbacks)
         tls_handler_offset = linker->getSymbolOffset("PETLSC2") + upxsection;
 
     processTls(&rel,&tlsiv,ic);
-    ODADDR(PEDIR_TLS) = sotls ? ic : 0;
-    ODSIZE(PEDIR_TLS) = sotls ? (sizeof(LEXX) == 4 ? 0x18 : 0x28) : 0;
-    ic = ALIGN_UP(ic + sotls, 4u);
+    ODADDR(PEDIR_TLS) = aligned_sotls ? ic : 0;
+    ODSIZE(PEDIR_TLS) = aligned_sotls ? (sizeof(LEXX) == 4 ? 0x18 : 0x28) : 0;
+    ic += aligned_sotls;
 
     processLoadConf(&rel, &loadconfiv, ic);
     ODADDR(PEDIR_LOADCONF) = soloadconf ? ic : 0;
@@ -2706,7 +2717,7 @@ void PeFile::pack0(OutputFile *fo, ht &ih, ht &oh,
         OutputFile::dump(opt->debug.dump_stub_loader, loader, codesize);
     if ((ic = fo->getBytesWritten() & (sizeof(LEXX) - 1)) != 0)
         fo->write(ibuf, sizeof(LEXX) - ic);
-    fo->write(otls,sotls);
+    fo->write(otls, aligned_sotls);
     fo->write(oloadconf, soloadconf);
     if ((ic = fo->getBytesWritten() & fam1) != 0)
         fo->write(ibuf,oh.filealign - ic);
@@ -2736,6 +2747,7 @@ void PeFile::pack0(OutputFile *fo, ht &ih, ht &oh,
     printf("%-13s: compressed   : %8ld bytes\n", getName(), (long) c_len);
     printf("%-13s: decompressor : %8ld bytes\n", getName(), (long) codesize);
     printf("%-13s: tls          : %8ld bytes\n", getName(), (long) sotls);
+    printf("%-13s: aligned_tls  : %8ld bytes\n", getName(), (long) aligned_sotls);
     printf("%-13s: resources    : %8ld bytes\n", getName(), (long) soresources);
     printf("%-13s: imports      : %8ld bytes\n", getName(), (long) soimpdlls);
     printf("%-13s: exports      : %8ld bytes\n", getName(), (long) soexport);
```
shows only the alignment changes by Laszlo, plus the extraneous addition:
```
+    if (mask & (
+             (1u<<IMAGE_SUBSYSTEM_EFI_APPLICATION)
+            |(1u<<IMAGE_SUBSYSTEM_EFI_BOOT_SERVICE_DRIVER)
+            |(1u<<IMAGE_SUBSYSTEM_EFI_RUNTIME_DRIVER)
+            |(1u<<IMAGE_SUBSYSTEM_EFI_ROM)
+              ))
+    {
+        infoWarning("PE: subsystem %u (EFI) lacks a run-time decompressor",
+            subsystem);
+    }
```
I will remove the extraneous addition.

Reply:
Thank you, @jreiser! I did some more git history checking and found that while it was added to the develop branch in f75ad8b, the merged jreiser-0522 branch had the same commit as 3d4d0d2 that is the most likely cause for the deleted lines being re-added. I don't think that there were any other merge issues, and I can confirm that after 23bc186 all the EFI changes are consistent with this PR.